--- 7z/C/Util/7z/7zMain.c
+++ 7z/C/Util/7z/7zMain.c
@@ -5,6 +5,7 @@

 #include <stdio.h>
 #include <string.h>
+#include <unistd.h>

 #include "../../CpuArch.h"

@@ -372,7 +373,7 @@
 
 // #define NUM_PARENTS_MAX 128
 
-int MY_CDECL main(int numargs, char *args[])
+int MY_CDECL lzma_main(int numargs, char *args[])
 {
   ISzAlloc allocImp;
   ISzAlloc allocTempImp;
@@ -390,7 +391,7 @@
   if (numargs == 1)
   {
     Print(
-      "Usage: 7zDec <command> <archive_name>\n\n"
+      "Usage: 7zDec <command> <archive_name> <location>\n\n"
       "<Commands>\n"
       "  e: Extract files from archive (without using directory names)\n"
       "  l: List contents of archive\n"
@@ -399,11 +400,12 @@
     return 0;
   }
 
-  if (numargs < 3)
+  if (numargs < 4)
   {
     PrintError("incorrect command");
     return 1;
   }
+  chdir(args[3]);
 
   #if defined(_WIN32) && !defined(USE_WINDOWS_FILE) && !defined(UNDER_CE)
   g_FileCodePage = AreFileApisANSI() ? CP_ACP : CP_OEMCP;
